src/internal/connector/exchange/service_restore.go

```diff
@@ -419,257 +202,145 @@ func restoreCollection(
 
-		folders := append([]string{dest}, directory.Folders()...)
+	for _, container := range containers {
+		pb = pb.Append(container)
 
-		return establishEventsRestoreLocation(
+		fid, err := getOrPopulateContainer(
 			ctx,
-			ac,
-			folders,
-			directoryCache,
-			user,
-			newCache,
+			cch,
+			cache,
+			pb,
+			userID,
+			container,
+			containerID,
+			isNewCache,
 			errs)
-
-	default:
-		return "", clues.New(fmt.Sprintf("type not supported: %T", category)).WithClues(ctx)
-	}
-}
-
-// establishMailRestoreLocation creates Mail folders in sequence
-// [root leaf1 leaf2] in a similar to a linked list.
-// @param folders is the desired path from the root to the container
-// that the items will be restored into
-// @param isNewCache identifies if the cache is created and not populated
-func establishMailRestoreLocation(
-	ctx context.Context,
-	ac api.Client,
-	folders []string,
-	mfc graph.ContainerResolver,
-	user string,
-	isNewCache bool,
-	errs *fault.Bus,
-) (string, error) {
-	// Process starts with the root folder in order to recreate
-	// the top-level folder with the same tactic
-	folderID := rootFolderAlias
-	pb := path.Builder{}
-
-	ctx = clues.Add(ctx, "is_new_cache", isNewCache)
-
-	for _, folder := range folders {
-		pb = *pb.Append(folder)
-
-		cached, ok := mfc.LocationInCache(pb.String())
-		if ok {
-			folderID = cached
-			continue
-		}
-
-		temp, err := ac.Mail().CreateMailFolderWithParent(ctx, user, folder, folderID)
 		if err != nil {
-			// Should only error if cache malfunctions or incorrect parameters
-			return "", err
+			return "", cache, clues.Stack(err)
 		}
 
-		folderID = ptr.Val(temp.GetId())
-
-		// Only populate the cache if we actually had to create it. Since we set
-		// newCache to false in this we'll only try to populate it once per function
-		// call even if we make a new cache.
-		if isNewCache {
-			if err := mfc.Populate(ctx, errs, rootFolderAlias); err != nil {
-				return "", clues.Wrap(err, "populating folder cache")
-			}
-
-			isNewCache = false
-		}
-
-		// NOOP if the folder is already in the cache.
-		if err = mfc.AddToCache(ctx, temp); err != nil {
-			return "", clues.Wrap(err, "adding folder to cache")
-		}
+		containerID = fid
 	}
 
-	return folderID, nil
+	return containerID, cache, nil
 }
 
-// establishContactsRestoreLocation creates Contact Folders in sequence
-// and updates the container resolver appropriately. Contact Folders are
-// displayed in a flat representation. Therefore, only the root can be populated and all content
-// must be restored into the root location.
-// @param folders is the list of intended folders from root to leaf (e.g. [root ...])
-// @param isNewCache bool representation of whether Populate function needs to be run
-func establishContactsRestoreLocation(
+func getOrPopulateContainer(
 	ctx context.Context,
-	ac api.Client,
-	folders []string,
-	cfc graph.ContainerResolver,
-	user string,
+	cch containerCacheHandler,
+	gcr graph.ContainerResolver,
```

Comment:
Made a pass on this implementation to simplify and highlight the differences.  This should clarify which interface is a client, and which is a stateful cache.

Reply:
>(containerCacheHandler, probably poorly named) handles graph api queries

yes, this needs a different name if it's not a cache. It's doing more than just handling queries, it's also the thing that is creating the container resolver which makes it feel reasonable to assume it may itself contain a container resolver if one doesn't dig through all of the code